executor/executor.go

```diff
@@ -1907,51 +1907,99 @@ func (e *TrimExec) Next() (*Row, error) {
 // UnionExec has multiple source Executors, it executes them sequentially, and do conversion to the same type
 // as source Executors may has different field type, we need to do conversion.
 type UnionExec struct {
-	schema expression.Schema
-	Srcs   []Executor
-	ctx    context.Context
-	cursor int
+	schema   expression.Schema
+	Srcs     []Executor
+	ctx      context.Context
+	inited   bool
+	finished atomic.Value
+	rowsCh   chan []*Row
+	rows     []*Row
+	cursor   int
+	wg       sync.WaitGroup
+	closedCh chan struct{}
+	errCh    chan error
 }
 
 // Schema implements the Executor Schema interface.
 func (e *UnionExec) Schema() expression.Schema {
 	return e.schema
 }
 
+func (e *UnionExec) waitAllFinished() {
+	e.wg.Wait()
+	close(e.rowsCh)
+	close(e.closedCh)
+}
+
+func (e *UnionExec) fetchData(idx int) {
+	defer func() {
+		e.wg.Done()
+	}()
+	for {
+		rows := make([]*Row, 0, batchSize)
+		for i := 0; i < batchSize; i++ {
+			if e.finished.Load().(bool) {
+				return
+			}
+			row, err := e.Srcs[idx].Next()
+			if err != nil {
+				e.finished.Store(true)
+				e.errCh <- err
+				return
+			}
+			if row == nil {
+				if len(rows) > 0 {
+					e.rowsCh <- rows
+				}
+				return
+			}
+			rows = append(rows, row)
+		}
+		e.rowsCh <- rows
+	}
+}
+
 // Next implements the Executor Next interface.
 func (e *UnionExec) Next() (*Row, error) {
-	for {
-		if e.cursor >= len(e.Srcs) {
-			return nil, nil
+	if !e.inited {
+		e.finished.Store(false)
+		e.rowsCh = make(chan []*Row, batchSize*len(e.Srcs))
+		e.errCh = make(chan error, len(e.Srcs))
+		e.closedCh = make(chan struct{}, 1)
+		for i := range e.Srcs {
+			e.wg.Add(1)
+			go e.fetchData(i)
+		}
+		go e.waitAllFinished()
+		e.inited = true
+	}
+	if e.cursor >= len(e.rows) {
+		var rows []*Row
+		var err error
+		select {
+		case rows, _ = <-e.rowsCh:
+		case err, _ = <-e.errCh:
```

Comment:
When we meet error in fetch data, we will set finished be true and notify all goroutines to stop. wg.Done is added in defer for every fetch data goroutines.